### ISSUE TITLE
Sample: Transfer a blob from a background context to a page

### ISSUE THREAD

**Issue by dotproto** [GoogleChrome/chrome-extensions-samples#766]
**Goal of the demo**

Demonstrate how an extension can create a blob in the extension's service worker and transfer it to a page's main world. 

**Suggested implementation**

Sketch of my initial implementation plan

1. Have a content script inject an iframe into a page
2. Run a script in the iframe to post a message back to the service worker using `navigator.serviceWorker.controller.postMessage()`
3. In the SW's message handler, generate a blob OR use `structuredClone()` to create a copy of a blob.
4. Use `event.source.postMessage(msg, [transferable])` to reply to the client and transfer the blob
5. In the iframe's message handler, use `window.parent.postMessage(msg, "<origin>", [transferable])` to transfer the data to the page

**Related links**

* https://github.com/w3c/webextensions/issues/293

**Notes**

Initial findings suggest that it's not possible to transfer variables across origins. I'm tentatively thinking that to work around this, we can use `URL.createObjectURL()` in the iframe and directly reference the object URL somewhere that's easily visible to the end user (e.g. canvas, Audio/Video element, etc.).

**Comment by guest271314** [comment]
Here you go https://github.com/guest271314/persistent-serviceworker/tree/main/chromium_extension_web_accessible_resources_iframe_message_event. You can transfer whatever you want. Bonus: Keeps MV3 `ServiceWorker` persistent.

**Comment by guest271314** [comment]
Note, the last time I checked `Blob`s are not transferable.

**Comment by guest271314** [comment]
Details of the message passing. We create a `MessageChannel` pair in the MV3 `ServiceWorker`, then transfer one of the `MessagePort`s created to the arbitrary Web page where action icon is clicked, to establish direct communication between `SeriveWorker` and the Web page where `port` is defined globally. 

To send a `Blob` from the `ServiceWorker` to the Web page where action icon is clicked include on Line 111 of background.js

```
port1.postMessage(new Blob(['Message from MV3 ServiceWorker'], {type:'text/plain'})); 
```
To log messages in the `ServiceWorker` sent from `port2` defined globally in the Web page where click action occurred add the following to Line 114 in background.js

```
  console.log(e.data);
```

To transfer substantial amounts of data we can transfer `ArrayBuffer` (and `TypedArray` `buffer`), `WebAssembly.Memory`, and on Chromium and Chrome Transferable Streams, e.g.,

```
const {readable,  writable} = new TransformStream();
port1.postMessage(readable, [readable]);
const writer = writable.getWriter();
await writer.write(new Uint8Array([1,2,3,...]));
// close the stream
await writer.close();
```
on Web page

```
port.onmessage = async(e) => {
  if (e.data instanceof ReadableStream) {
    await readable.pipeThrough(...).pipeTo(...);
  }
}
```

**Comment by tophf** [comment]
> Goal of the demo
> transfer it to a page's main world

It's the wrong goal in the context of the linked issue. I guess its title was initially confusing as it mentioned "web page" instead of "web page's content script", so I've change it to reflect the _contents_ of the issue, which is about the content script. It's a more difficult task to accomplish securely, but still possible via MessagePort as described in the [comment linked in the issue](https://github.com/w3c/webextensions/issues/77#issuecomment-1243308962). One change to the scheme would be to send the MessageChannel port from SW to the content script via the iframe, then use this port to transfer blobs instantly by specifying the second parameter of postMessage.

**Comment by guest271314** [comment]
> One change to the scheme would be to send the MessageChannel port from SW to the content script via the iframe

That is what the code I posted does.

> Note that to transfer a huge blob instantly, which is what the related issue is all about, we need the second parameter:
>
> `port1.postMessage(blob, [await blob.arrayBuffer()]);`

As far as I know `Blob`s are not transferable and that code does not transfer a `Blob`.

**Comment by tophf** [comment]
Blob is just a wrapper on the same data exposed via arrayBuffer accessor or other getters on the blob, so this code does transfer the actual data instantly.

**Comment by guest271314** [comment]
Kindly cite where you read `Blob`s are transferable.

**Comment by tophf** [comment]
Not the Blob, the arrayBuffer view of the same data contained in the Blob.

**Comment by guest271314** [comment]
That is not transferring a `Blob`. Just deal with `ArrayBuffer` and `TypedArray` if you want to transfer data.

**Comment by tophf** [comment]
A Blob will be received by the listener, so it can be said that the Blob is effectively transferred.

**Comment by guest271314** [comment]
No it cannot be said, because `Blob`s are not transferable objects.

**Comment by tophf** [comment]
Blob is not a `Transferable` but it's efectively transferable. Anyway, it's semantics.

**Comment by guest271314** [comment]
Note that `Blob` is not listed here https://developer.mozilla.org/en-US/docs/Glossary/Transferable_objects#supported_objects.

**Comment by tophf** [comment]
I know.

**Comment by guest271314** [comment]
Then as I said you cannot transfer a `Blob`. 

What you can do is pass a `FileSystemHandle` to the `ServiceWorker` from the main Web page and write data to the `FileSystemFileHandle` then transfer the object back to the main frame then call `.getFile()`.

**Comment by guest271314** [comment]
Setting the type of data and object that can be and cannot be transfered aside for the moment, the code I posted achieves the requirement.

**Comment by tophf** [comment]
Using the code I posted (with the second parameter using blob.arrayBuffer) will transfer a Blob instantly, it's trivial to verify.

**Comment by tophf** [comment]
FWIW, if you're interested what happens under the hood, here's my guess: the blob's actual data being a Transferable (blob.arrayBuffer) is sent instantly, while the Blob itself is just a wrapper over that data, which is structuredClone'able, so it's reconstructed in the receiver using the instantly transferred data.

**Comment by guest271314** [comment]
Where did you learn that calling `Blob.arrayBuffer()` at 2d parameter of `postMessage()` will

> transfer a Blob instantly

?

In what specification is that written? Kindly cite your sources.

What happens when you post the `Blob` without 2d parameter to `postMessage()`?

**Comment by tophf** [comment]
1. It's how Blob and other binary related wrappers (DataView, typed arrays, buffer) work. You can investigate it yourself.
2. The blob will be copied via internal structuredClone.

**Comment by guest271314** [comment]
You haven't answered the question I asked re source of your claim?

**Comment by tophf** [comment]
It's in the specifications somewhere, which I investigated when I looked for a way to transfer blobs instantly. I found it, confirmed it worked. I don't have the links in browser history anymore, so if you're interested you can investigate and verify it yourself.

**Comment by guest271314** [comment]
No, it is not in the specifications.

And it doesn't happen https://plnkr.co/edit/ZyutxttrhgerqZpk?open=lib%2Fscript.js.

**Comment by guest271314** [comment]
Anyway, I think the trasferable `Blob` claim can be set aside.

The Web site removing the `iframe` of a Chrome extension is more interesting and far more difficult to prove by reproduction.

**Comment by tophf** [comment]
Your demo prints the blob in console, just as it should.

**Comment by guest271314** [comment]
The `Blob` is not transfered whatsoever. Note both the `Blob` in the Worker and the `Blob` received on main page have `size` 3.

That means 

```
[await blob.arrayBuffer()]
```

is completely superfluous.

**Comment by tophf** [comment]
Maybe there's no need for arrayBuffer, but try making a 1GB Blob and you will see it's transferred instantly, which is the only thing I care about in this context.

**Comment by guest271314** [comment]
You are using incorrect terminology The `Blob` is not _transfered_ at all. You will have 1GB `Blob` in `DedicatedWorker` or `ServiceWorker` contexts _and_ in main thread.

**Comment by tophf** [comment]
Try sending `['a'.repeat(100e6)]` and `new Blob(['a'.repeat(500e6)])`. The string will take 1sec, the blob will be instantly sent.

**Comment by guest271314** [comment]
As long as you have ceased and desisted using the term _transferred_ we are making progress.

I won't go in to the "instantly" claim as that is essentially the same as transferred. 

`Blob` construction is expensie, and not instant.

If you want instant you should use `SharedArrayBuffer`.

**Comment by guest271314** [comment]
We have established that we can post `Blob`s from the `ServiceWorker` to the arbitrary Web page.

**Comment by tophf** [comment]
Okay, I see now where I was wrong. Apparently the speed difference confused me. The Blob is sent instantly even without being actually *transferred*, whereas a string takes a lot of time probably due to the need to intern it in the receiver.

**Comment by guest271314** [comment]
No. That is not what is occurring.

To learn about `Blob`s in Chrome I suggest reading this answer https://stackoverflow.com/a/56419176 at [Where is Blob binary data stored?](https://stackoverflow.com/questions/38239361/where-is-blob-binary-data-stored).

**Comment by tophf** [comment]
It means that to achieve the true instant transfer, which is the goal of the linked issue, the demo should send `[await blob.arrayBuffer(), blob.type]`, and the receiver's listener should reconstruct the Blob from this data. I see this was correctly done in the [original post](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/IPJSfjNSgh8/m/Dh35-tZPAgAJ) linked in the comment linked in the issue, which I incorrectly reproduced from my faulty memory, sorry.

**Comment by guest271314** [comment]
Alright. That example is basically the code I posted here https://github.com/guest271314/persistent-serviceworker/tree/main/chromium_extension_web_accessible_resources_iframe_message_event which I don;t think you have tested yet. That is possible. You are just posting an `Array` to the main page.

**Comment by guest271314** [comment]
As I commented above, File System Access API `FileSystemFileHandle` _is_ transferable. Which means you can pass a file handle to the `ServiceWorker`, write data using `FileSystemWritableFileStream` and if necessary post (_transfer_) that same `FileSystemFileHandle` back to the main browsing context - or write the file directly to users' file system.

I am not sure why you focus is on a `Blob`? 

Yet, you can follow the instructions I posted above https://github.com/GoogleChrome/chrome-extensions-samples/issues/766#issuecomment-1294267869 to post a `Blob` from MV3 `ServiceWorker` to arbitrary Web page using `MessageChannel`.

**Comment by tophf** [comment]
A Blob is just an example of a popular binary type, nothing special.
It might be useful for this demo to show how to transfer a few other binary types as well, including streams.

**Comment by guest271314** [comment]
I did show that. I transfer a `ReadableStream` to the arbitrary Web page with `postMessage()` and read the data using Streams API.

**Comment by guest271314** [comment]
The code in this comment https://github.com/GoogleChrome/chrome-extensions-samples/issues/766#issuecomment-1294267869 streams raw PCM at ~10MB per minute. Note also that given the Native Messaging host use the source data processed in `onMessage` handler can also be JSON in the form of `"[0.123, 0.456, ...]"` which is then converted to a JavaScript object then passed to `Uint8Array` where that data is streamed to the browser.

**Comment by guest271314** [comment]
@tophf What are you actually trying to do?

**Comment by tophf** [comment]
Nothing specific personally. The linked issue simply summarizes the need previously expressed by other extension authors who aren't present in this group.

**Comment by guest271314** [comment]
W3C banned me, so I am unable to post on W3C or WICG repositories.

As demonstrated in the linked repository code it is possible to use `postMessage()` to send data from an MV3 `ServiceWorker` to an arbitrary Web page, and vice versa.

**Comment by guest271314** [comment]
Re the unverified conjecture that a We page could run code constantly to check to determine if a Web page has loaded an `iframe` with `src` set to `chrome-extension:` protocol and then remove the `iframe`, to avoid that concern that so far has not been reproduced, simply open the HTML page listed in `"web_accessible_resources"` as a top-level window in a Tab, then that hypothetical concern is mitigated.

**Comment by guest271314** [comment]
@tophf I will reiterate here that the approach, while possible, is not as ergonomic as it could be.

Ideally we can directly assign an arbitrary Web page as a `WindowClient` https://developer.mozilla.org/en-US/docs/Web/API/WindowClient, then we will have direct communication without an `iframe`.

Nonetheless the requirement is possible right now. 

I have streamed over an hour of raw PCM through the embedded `iframe` multiple times using the above approach. That is well over 650MB.

**Comment by tophf** [comment]
To be more precise, a site can't read or see an iframe inside closed ShadowDOM as it won't be exposed as `window[0]` or the identical `frames[0]` (at least in Chrome; FF incorrectly exposes it). FWIW, even without using ShadowDOM the site won't see the exact URL of the iframe if we create a src-less iframe, then set its frameElem.contentWindow.location.href to web_accessible_resources URL, and the web site will never be able to read the actual URL, it will only be able to infer that the iframe is cross-origin. Still, it's physically possible for a site to delete all DOM elements that it didn't create as it's pretty straightforward to implement.

Note that the linked issue is about secure messaging to a content script specifically, not a web page, so any demo that deals with the web page's MAIN world is not really related to that issue.

**Comment by guest271314** [comment]
> Still, it's physically possible for a site to delete all DOM elements that it didn't create as it's pretty straightforward to implement.

Kindly cite an example of that occurring in the wild.

> Note that the linked issue is about secure messaging to a content script specifically, not a web page, so any demo that deals with the web page's MAIN world is not really related to that issue.

A content script is just code injected into `globalThis`. `chrome.scripting.executeScript()` pulls back the veil on that.

There is no such thing as "secure" in the domain of signal communications.

**Comment by tophf** [comment]
It doesn't matter how many times you ask me to cite a URL, I won't do it. Regardless of our personal experience, it remains physically possible. There are extensions that can't accept that risk.

A content script runs in the isolated world which cannot be accessed and intercepted by the web page's MAIN world (unless there's a bug in the browser), which is what makes such communication secure. Since we're not writing a scientific article this term is apt.

**Comment by guest271314** [comment]
I think you have a misconception of what "secure" means.

As of last century certain entities were capturing and analyzing 20TB of Internet data in real-time https://agoodamerican.org/. There is no way for you to verify your data has not been intercepted.

I don't think it is wise to throw around terms like "secure" without understanding the physical topography of the medium.

There are scientific terms as we are discussing technical writing. 

That is, if the requirement is not possible, then somebody has to write out what they intend to implement, using technical writing to cite prior art, goals, non-goals, etc.

Defining "secure" is impossible, as it doesn't exist.

**Comment by tophf** [comment]
Any term is fine by me as long as it means "cannot be accessed and intercepted by the web page's MAIN world unless there's a bug in the browser".

**Comment by guest271314** [comment]
> It doesn't matter how many times you ask me to cite a URL, I won't do it. 

You can't.

The claim is pure conjecture.

> Any term is fine by me as long as it means "cannot be accessed and intercepted by the web page's MAIN world unless there's a bug in the browser".

You can't guarantee that.

Notably that counters your claim that Web sites will remove `iframe`s from their site. How would they do that when the I `iframe` is only appended on your machine, using your files on your machine?

**Comment by tophf** [comment]
Well, I can physically make such a site myself, so I physically can cite a URL. It's just meaningless because it won't change the fact that it's physically possible. There are sites who don't like extensions e.g. those that can help the visitors cheat if it's a game or a test or something more sensitive, so they do have an incentive to do extra stuff to fight extensions.

As for the sites dig, I guess you're trying to make a joke about my imprecise usage of the term "site", but it's widely used (I preemptively decline a citation request) to denote the contents of a local browser's tab, so naturally it can remove anything from the tab's DOM where the web page of the web site is shown.

**Comment by guest271314** [comment]
> Well, I can physically make such a site myself, so I physically can cite a URL.

No, you can't.

You will need to demonstrate how you will make a Web site which will remove the `<iframe>` that I append to the `document` I am viewing in my browser on my machine.

**Comment by guest271314** [comment]
> Well, I can physically make such a site myself, so I physically can cite a URL.

Further, per your claim the Web site can remove any code whatsoever, HTML, CSS, JavaScript, etc., whatever you consider to be a "secure" implementation of communication - including `chrome` itself 

```
chrome = null;
```
Now what do you do?

**Comment by tophf** [comment]
> No, you can't.

I can.
You seem to keep milking the joke about differentiating remote sites and their local rendition in a browser tab.
Here's the simplest demo of such a web site html:
```html
<script>addEventListener('DOMNodeInserted', e => e.target.remove())</script>
```

A more "advanced" version would also remove the stuff added by a content script at document_start:
```html
<script>
document.documentElement.querySelectorAll('*').forEach(el => {
  if (el !== document.head) el.remove();
});
addEventListener('DOMNodeInserted', e => e.target.remove());
</script>
```

Next the site can show its own elements and keep track of them in an internal WeakSet.

>chrome = null;
Now what do you do?

Assuming it's in the web page's MAIN world, it doesn't affect the isolated world of the content scripts.

**Comment by guest271314** [comment]
Well, I can just remove that code.

**Comment by tophf** [comment]
It will be obfuscated in the real web page and it'll contain the main functionality of the page, so removing it is not an option. What an extension can do is to preemptively cloak the various DOM methods.

**Comment by guest271314** [comment]
> It will be obfuscated in the real web page and it'll contain the main functionality of the page, so removing it is not an option. What an extension can do is to preemptively cloak the various DOM methods.

No it won't. 

There is no obfuscation for hackers.

**Comment by tophf** [comment]
It will be interspersed with the real code of the web page that has already been executed, so removing the script element after the fact won't help. Cloaking the DOM methods would help, but there's a lot to cover, especially if we include the possibility of a web site using a fresh src-less iframe to extract the original un-cloacked methods and .call them on the main document. Anyway, I agree it's not something most extensions need to worry about.

**Comment by guest271314** [comment]
I think it is important to point out while the process I use to stream data does work, it could be written out to avoid using an `iframe`, and be ergonomic and overt to the developer.

Technically you can just `fetch()` whatever data you want using `fetch('chrome-extension://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'')`.

**Comment by guest271314** [comment]
You have to be real pecise when dealing with technical documents. E.g., Chrome claims to capture "system audio" for `getDisplayMedia()`, yet there are exceptions which I requested be disclosed to the developers at large, yet the reply was a "small percentage" of users would even be aware of `speechSynthesis` https://github.com/GoogleChrome/developer.chrome.com/pull/3947 - until you try to capture `speechSynthesis.speak()` yourself for a requirement.

So, I would be very technical about exactly what is being requested here for a feature - with use cases and non-goals clearly spelled out so confusion as to scope cannot rationally be claimed.

**Comment by dotproto** [comment]
Dang, y'all got into quite a bit of a discussion. I don't have time to catch up on it right now, but I swung by to add some notes and figured I should address one point that stood out in passing.

> > Goal of the demo
> > transfer it to a page's main world
> 
> It's the wrong goal in the context of the linked issue. 

I was initially aiming to tackle what I expected to be the more complicated version of the same basic use case; I figured if we could solve for passing a Blob to a main world content script, then we'd probably also solve for passing a Blob to an isolated world content script as well.

Unfortunately, my initial experiments seem to indicate that transferables cannot be sent across origin. At the moment my planned work around is to create object URL in the extension iframe on the page. That may change after I read through the full discussion in the comments here.

**Comment by guest271314** [comment]
> my initial experiments seem to indicate that transferables cannot be sent across origin

Yes they can be, if we are talking about from `chrome://<id>` to arbitrary Web page, and vice versa.

**Comment by guest271314** [comment]
@tophf 

Reviewing my own approach, I don;t think we have to keep the `iframe` in the Web page once we _transfer_ the `ReadableStream` to the Web page. So the `iframe` would only serve to transfer the `ReadableStream` to the Web page then we can remove it ourselves, which should take 1 second or so.

I'll test later today or tomorrow.

**Comment by tophf** [comment]
I would just transfer port2 of a MessageChannel as you already suggested, so that it can be also used for future communication.

**Comment by guest271314** [comment]
@tophf Is there any reason you don't use `"externally_connectable"`?

**Comment by dotproto** [comment]
> > my initial experiments seem to indicate that transferables cannot be sent across origin
> 
> Yes they can be, if we are talking about from `chrome://<id>` to arbitrary Web page, and vice versa.

Thanks for the clarification. I was experimenting with websites, not extension contexts. I should have known better 😅

**Comment by guest271314** [comment]
> The linked issue is about an entirely different use case: a content script. Content scripts run in an isolated world which cannot be spoofed or intercepted or broken by the web page via getters/setters on various global types and prototypes.

Doesn't really matter. You can stream data using message passing. There is nothing special about a `Blob`. JSON can be streamed. Does "content script" exclude `chrome.scripting.executeScript()`?

**Comment by tophf** [comment]
You asked about externally_connectable, which is about the opposite task: collaborating with the main world of the page.

**Comment by guest271314** [comment]
I am trying to understand exactly what you are trying to achieve that you can't now.

In the content script you can just do

```
chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
  // do stuff
});
```

in the `ServiceWorker`

```
chrome.action.onClicked.addListener(async (tab) => {
  const [{ id }] = await chrome.tabs.query({
                           active: true,
                           currentWindow: true,
                        });
  chrome.tabs.sendMessage(id, { name: 'blob' }, async (message) => {
    console.log(message);
  });    
});
```

We can serialize data of you really have a `Blob` or `TypedArray` using `TextEncoder()` and deserialize with `TextDecoder()`.

**Comment by tophf** [comment]
The linked issue is about transferring huge amounts of data "instantly", i.e. in a matter of a few milliseconds, while staying wholly within the isolated world so that nothing in the main world can see the data or break the communication.

It means that current `chrome` messaging is immediately out of the question because it would take like a minute to transfer 1GB of JSON-ified binary data on an average not-so-fast notebook.

The workaround via a web_accessible_resources iframe is practically acceptable for most extensions that can accept the theoretical risks of being detected/broken from the main world and the added 50+ms overhead to create the iframe environment.

**Comment by guest271314** [comment]
> The linked issue is about transferring huge amounts of data "instantly", i.e. in a matter of a few milliseconds, while staying wholly within the isolated world so that nothing in the main world can see the data or break the communication.
>
> It means that current chrome messaging is immediately out of the question because it would take like a minute to transfer 1GB of JSON-compatible binary data on an average not-so-fast notebook.

Are you sure you tested that?

What data are you transferring from the `ServiceWorker` to the content script where the transfer _needs_ to be 1GB or more in under 1 second and instantaneous?

What is the use case?

I can stream real-time audio using data URL's or JSON with `onMessage`. Streaming audio ain't easy. It is easy to detech glitches, gaps in playback.

What I would do is _transfer_ a `FileSystemFileHandle` to the `ServiceWorker`, write to the file, then either close the stream or transfer the `FileSystemFileHandle` back to the content script, call `getFile()` then call `arrayBuffer()` on the `File` object.

Or _transfer_ a `WebAssembly.Memory` object and write and read essentially in parallel.

I am not sure why the context restriction is content scripts, which are inherently unreliable.

For more reliability you can just open the document listed in `"web_accessible_resources"` as a Tab, then use `fetch()`, intercept the request and send data back using `respondWith()`.

The

> on an average not-so-fast notebook

criteria will remain the same using any approach.

**Comment by guest271314** [comment]
@dotproto

Changing the name to "Sample: Transfer a blob from a background context to a page" is still not techically accurate, because `Blob`s are not transferable objects. The last time I checked Chrome still had not implemented `Blob` type for WebRTC data channels.

The sample request could be solved at once by various means:

1. Use the origins set, if any, in `"web_accessible_resources"` and apply those to allowlist `BroadcastChannel` from chrome-extension://<id> origin;
2.  Use the origins set, if any, in `"web_accessible_resources"` and apply those to allowlist the content script as a `WindowClient` of MV3 `ServiceWorker`

Either will provide a means to communicate between the content script/Web page and the `ServiceWorker`.

A content script as `WindowClient` will alos allow intercepting `fetch()` requests so we can implement half-duplex streaming.

The `BroadcastChannel` solution appears to be the simplest to implement in extension source code, as `ServiceWorker` was not designed for extensions.

**Comment by guest271314** [comment]
> The linked issue is about transferring huge amounts of data "instantly", i.e. in a matter of a few milliseconds, while staying wholly within the isolated world so that nothing in the main world can see the data or break the communication.
> The workaround via a web_accessible_resources iframe is practically acceptable for most extensions that can accept the theoretical risks of being detected/broken from the main world and the added 50+ms overhead to create the iframe environment. To clarify "detection": assuming we use a closed ShadowDOM the web page can only detect that an empty element was added, whereas a content script of another extension can peek into the closed ShadowDOM and see there's an iframe (it won't be able to peek into iframe though without resorting to chrome.debugger API which is an entirely different story).

There is nothing in developer.chrome that specifies a means to _transfer_ data between MV3 `ServiceWorker` and content script, or Web page contexts.

`Blob`s are not transferable.

So this issue amounts to a feature request.

If the concern is an arbitrary Web page running code continuously to check check and see if an `iframe` is appended to the `document`, which I have not observed in the wild whatsoever, you can open a new `window`, which the Web page has no control over observing or closing and achieve the same transfer of data between `iframe` and Web page.

Something like I do here https://github.com/guest271314/sw-extension-audio/blob/main/background.js

```
const createAudioWindow = async (notification = false) => {
  let url = chrome.runtime.getURL('audio.html');
  if (notification) {
    url += '?notification=1';
  }
  console.log(url);
  ({ id } = await chrome.windows.create({
    type: 'popup',
    focused: false,
    top: 1,
    left: 1,
    height: 1,
    width: 1,
    url,
  }));
  await chrome.windows.update(id, { focused: false });
  return id;
};
```

Now you can communicate using `opener.postMessage()` from the `window` and `onmessage` in the Web page, then close the `window` when done with the transfer.

**Comment by guest271314** [comment]
Evidently `setSelfAsOpener` option of `chrome.windows.create()` doesn't work with `ServiceWorker`, even though it is not documented to throw.

Per this https://stackoverflow.com/a/72124984 Chrome maximum `ArrayBuffer` size is 2145386496 (1.998 GB).

`new ArrayBuffer(1072693248)` 

half of the maximum value per that test crashed the `window` on *nix. 

`WebAssembly.Memory` can `grow()`, the last time I checked was limited as well, though there was activity to increase to 4GB https://v8.dev/blog/4gb-wasm-memory,  https://github.com/WebAssembly/spec/issues/1116.

`new Array(536346624)`

.499 GB did not throw using the following approach; was not instant. You should be able to run the test linked above between 536346624 and 1072693248 to determine when Chrome or Chromium will crash on your machine. The sole purpose is to transfer data from `ServiceWorker` to content script (or Web page) then close the `window`

manifest.json

```
{
  "name": "ServiceWorker => ArrayBuffer => Content script",
  "version": "1.0",
  "description": "Transfer data from ServiceWorker to content script",
  "manifest_version": 3,
  "background": {
    "service_worker": "background.js",
    "type": "module"
  },
  "content_scripts": [
    {
      "matches": [
        "<all_urls>"
      ],
      "js": [
        "contentscript.js"
      ],
      "run_at": "document_start"
    }
  ],
  "host_permissions": [
    "http://*/*",
    "https://*/*"
  ],
  "permissions": [
    "activeTab",
    "tabs",
    "windows"
  ],
  "web_accessible_resources": [{
      "resources": ["*.html", "*.js"],
      "matches": ["<all_urls>"],
      "extensions": []
  }]
}
```

background.js

```
const bc = new BroadcastChannel('transfer');
const buffer = new ArrayBuffer(536346624); 
bc.onmessage = async (e) => {
  console.log(e);
  const transfer = structuredClone(buffer, {transfer: [buffer]});
  bc.postMessage(transfer, [transfer]);
  console.assert(buffer.byteLength === 0, {buffer});
}

chrome.action.onClicked.addListener(async (tab) => {
  chrome.tabs.sendMessage(tab.id, { url: chrome.runtime.getURL('index.html') });    
});
```

contentscript.js

```
{
  chrome.runtime.onMessage.addListener((message) => {
    const handleMessage = (e) => {
      if (e.origin === new URL(message.url).origin) {
        console.log(e.data);
        removeEventListener('message', handleMessage);
      }
    }
    addEventListener('message', handleMessage);
    window.open(message.url, location.href, 'menubar=no,location=no,resizable=no,scrollbars=no,status=no,width=50,height=50');
  });
}
```

index.html

```
<script src="./script.js"></script>
```

script.js

```
const bc = new BroadcastChannel('transfer');
bc.onmessage = (e) => {
  opener.postMessage(e.data, name, [e.data]);
  bc.close();
  close();
}
bc.postMessage(null);
```

**Comment by tophf** [comment]
The linked issue is indeed a feature request to implement the means to transfer huge data "instantly". I've edited it to avoid the fixation on blobs as it's not about blobs in particular, but any binary data.

I've retested the actual time:

* Sending a 500MB blob: ~1.5sec
* Transferring a 500MB blob's ArrayBuffer: ~0.5sec

There are many use cases where huge binary data needs to be sent "instantly", for which `chrome` sendMessage is 10-100 times slower because it applies JSON.stringify that escapes some characters and encodes the result to UTF-8 again, so anything non-ASCII you provide to sendMessage, including text you just manually encoded to UTF-8 or any other at least 8-bit encoding, will be re-encoded slowly again. Then the reverse will be applied in the receiver.

It means that `chrome` sendMessage stresses the CPU excessively, which can be observed in profiler and task manager. It blocks the extension process and the receiver process and sometimes the browser process as well, introducing lags and jank.

> I am not sure why the context restriction is content scripts, which are inherently unreliable.

It's the opposite. Anything outside the content script is unreliable, i.e. the main world of a web page.

**Comment by guest271314** [comment]
How did you send a `Blob`? A `Blob` is not transferable. None of these transfers occur "instantly".

I was comparing `chrome.scripting.executeScript()` to a content script.

You have not described a use case where speed of transfer is important or required.

You cannot spoof, intercept of otherwise do anything about the content script or Web page itself opening a new window.

**Comment by tophf** [comment]
I've used different terms specifically to indicate the difference between sending and transferring. Sending is 
`port.postMessage(blob)`, transferring is `port.postMessage(buf, [buf])`. My measurements for sending a blob in the comment above were incorrect as I've included the time to fill the blob with random values. The actual time is less than 1ms i.e. practically instantaneous. This was what confused me the first time I investigated transferability of blobs. Apparently blobs are using a shared repository, you've mentioned something like that, so when we send a blob only the handle is actually sent. Time to transfer a 500MB buffer was correct, it's 0.5sec;

chrome.scripting.executeScript() injects a content script by default unless you specify `world: 'MAIN'` and the injected script becomes a "page script", not a "content script". I name it a "page script" because it runs in the JS environment of the page, so it plays by the same rules, the only difference being that it's not subjected to the CSP of the page.

Speed of transfer is important or required depending on the amount of data and whether sending it introduces lags or stresses the CPU so much that its fan becomes audible on a low-power device like notebooks. Most extensions don't *need* this feature. In your case the difference might be negligible, otherwise you would have probably noticed these problems.

Opening a new window is a viable workaround, however depending on the desktop environment and its settings it may introduce noticeable movement in some area of the screen and it may be also closed by another extension before the script in the window initializes.

**Comment by guest271314** [comment]
> Speed of transfer is important or required depending on the amount of data and whether sending it introduces lags or stresses the CPU so much that its fan becomes audible on a low-power device like notebooks. 

I am still not certain what use case `onMessage` doesn't work with.

Why would we need to _transfer_ 500MB of data to the content script? 

If the use case is streaming media, `onMessage` works.

If the use case is downloading files we use `chrome.downloads` or File System Access API on the main thread - or we can _transfer_ the `FileSystemFileHandle` to the `ServiceWorker` and write to the user file system there.

If you really want real-time communication you can use Native Messaging with libdatachannel as part of the host, so you can create a data channel connection from the content script.

**Comment by guest271314** [comment]
> Opening a new window is a viable workaround, however depending on the desktop environment and its settings it may introduce noticeable movement in some area of the screen and it may be also closed by another extension before the script in the window initializes.

The idea is not to occlude the `window`, rather make the window obvious so the user knows that window is only being used to transfer data to the content script.

Another extension can close any window or tab at any time, anyway.

**Comment by guest271314** [comment]
> Time to transfer a 500MB buffer was correct, it's 0.5sec;

Where does that data come from?

You can just use `fetch('chrome-extension:/path/to/file')` in the content script, then you have the speed of `Fetch` and `ReadableStream` as response which you can read while the stream is still in progress.

**Comment by guest271314** [comment]
I think we can write the data directly with `executeScript()` to avoid having to transfer data

```
chrome.action.onClicked.addListener(async (tab) => {
  const [{ result }] = await chrome.scripting.executeScript({
    target: {
      tabId: tab.id,
    },
    world: 'ISOLATED',
    func: () => { 
      const buffer = new ArrayBuffer(2145386496 / 2); 
      console.log(buffer); 
      return buffer.byteLength;
    }
  });
  
  console.log(result);

});
```

**Comment by tophf** [comment]
Only func's code is sent as a string to the tab where a new function is created from this text, not the data. You can send the data via `args` parameter but internally that's the same as sendMessage because it also uses JSON.stringify + IPC. This API is just a polyfill for MV2 executeScript with ```code: `(${func})(${JSON.stringify(args).slice(1,-1)})```.

I found a workaround that may be unbreakable:

1. create an offscreen document in SW
2. create an iframe there with the URL (or just the origin) of the web page in the tab
3. call `window.stop()` in your content script that runs in that iframe at document_start, it's temporarily registered for the duration of the whole thing using chrome.scripting API, it should check location.ancestorOrigins to see whether it's embedded inside an extension document.
4. SW sends a random id to the content script in the tab and to the offscreen iframe's content script
5. these content scripts open `BroadcastChannel` using this id as a name
6. the offscreen iframe sends a `Blob`, which takes less than 1 millisecond

It may need setting up declarativeNetRequest rule to [strip X-Frame-Options:DENY](https://stackoverflow.com/questions/15532791/getting-around-x-frame-options-deny-in-a-chrome-extension) and another rule to strip `referer` with the extension's id of the embedder origin if it's sent.

Currently I don't know how to obtain the `window` of the tab via this approach in order to be able to *transfer* stuff like a MessagePort. Might be impossible.

Another disadvantage is a possibly long time to open a slow server URL.

**Comment by guest271314** [comment]
> Only func's code is sent as a string to the tab where a new function is created from this text, not the data.

What difference does it make?

> 1. create an offscreen document in SW

What is that? Something Chromium extension authors have been talking about?

> Currently I don't know how to obtain the window of the tab via this approach 

Not sure what you mean.

**Comment by tophf** [comment]
The difference is that your code doesn't send any data, it just sends func.toString().

**Comment by guest271314** [comment]
That is not observble. You get the same data.

**Comment by guest271314** [comment]
What does the data in the `Blob` consist of? What is the use case?

**Comment by tophf** [comment]
No, you don't get any data at all. The data is not a part of the function's code string.

**Comment by tophf** [comment]
The data is any binary generated or downloaded stuff. It is not a part of any function.

**Comment by guest271314** [comment]
It doesn't matter. In the clicked `window` all you know is there is you `Blob`,

**Comment by tophf** [comment]
It won't contain data from SW, it's a useless exercise in semantics.

**Comment by guest271314** [comment]
I don't think it matters. You create the data structure you want to be exposed in the `window` in the `ServiceWorker`. The clicked `window` gets that data. 

So far you have not described exactly what the data consists of, and why it is important to _transfer_ the data.

**Comment by guest271314** [comment]
Is there any reason you don't just use `fetch('chrome-extension://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx')` from the arbitrary `window` to get the data?

**Comment by guest271314** [comment]
I was giving the idea of an arbitrary Web site removing an `iframe` from their site yesterday. 

If that happens once, then you can simply remove that code from the site after your initial test.

Given the challenge of writing source code that removes any arbitrary `iframe` from a Web site _and_ prevents said source code from being removed, such a requirement would be essentially impossible.

**Comment by guest271314** [comment]
If you already have the data that needs to be transferred you can just set `"web_accessible_resources"` in the manifest.json then use `fetch('chrome-extension://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx/path/to/file')` or `import json from 'chrome-extension://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx/path/to/file' assert { type: "json" }` on the Web site itself.

**Comment by tophf** [comment]
We're going in circles for like 100th time, but it's fun so let's do it once again.

A web page can trivially delete any element immediately once it's added as I've shown using DOMNodeInserted event. It can be done using MutationObserver too, as well as using a periodic check, it doesn't matter which one as they all are trivial to implement. It can't be removed automatically from an unknown web page not because it specifically resists removal but because such code will be an integral part of a large minified bundle, possibly with obfuscation - you'd have to analyze it manually. Such code can be added by another extension as well, not that it's likely, but still possible and trivial to implement. It is trivial, there's no challenge in writing it. Of course there's a way for an extension to protect the iframe: it can override all DOM methods that can be used to remove an element or to rewrite/reload the page or to extract such methods from a new src-less iframe.

> executeScript
> fetch
> import

All of these are for static data included with the extension, it's not what the linked issue is about. The issue is about data in SW context. It can be generated by JS or downloaded from an external URL or captured from desktop. It's not a part of the func's code or of the extension package.

**Comment by guest271314** [comment]
> This code can't be removed automatically from an unknown web page not because it specifically resists removal but because such code will be an integral part of a large minified bundle, possibly with obfuscation - you'd have to analyze it manually. 

The code can be removed. If need be we can use Local Modifications to run the local code of our choosing.

> The issue is about data in SW context (and nowhere else within the extension). It can be generated by JS or downloaded from an external URL or captured from desktop. It's not a part of the func's code or of the extension package.

However the code is generated it can be written to a local file and fetched with `fetch()` or `import()`, or `import`.

**Comment by guest271314** [comment]
>  you'd have to analyze it manually. 

Yes, one time. At the initial visit to any Web site that happens to runs such code, which I have never encountered in the field.

Thereafter we don't have that issue ever again.

> All of these are for static data included with the extension, it's not what the linked issue is about. The issue is about data in SW context (and nowhere else within the extension). It can be generated by JS or downloaded from an external URL or captured from desktop. It's not a part of the func's code or of the extension package.

The download and capture from desktop use cases can be excluded altogether.

If we are really downloading content in the `ServiceWorker` we don't need to transfer that data to the content script at all. Why would we do that? We just write the data directly to the user file system with `FileSystemFileHandle`. If we are capturing content from the desktop, again, why would be send that data to the `ServiceWorker` then back to the content script?

> A web page can trivially delete any element immediately once it's added as I've shown using DOMNodeInserted event. It can be done using MutationObserver too, as well as using a periodic check, it doesn't matter which one as they all are trivial to implement. This code can't be removed automatically from an unknown web page not because it specifically resists removal but because such code will be an integral part of a large minified bundle, possibly with obfuscation - you'd have to analyze it manually. Such code can be added by another extension as well, not that it's likely, but still possible and trivial to implement. It is trivial, there's no challenge in writing it. 

I really don't think you can write such code that cannot be removed. In absense of an actual proof-of-concept of such code working in the field I don;t think you have demonstrated the feasibility of someone even trying to do that.

**Comment by tophf** [comment]
Although Local Modifications is not a part of the extensions API/platform, but indeed its functionality can be replicated via chrome.debugger API, and indeed it can be used to remove the code once we identified it manually. I didn't contest this part. My point is that the analysis for such code can't be reliably automated. Also, chrome.debugger is not an API a typical user would want to allow for an extension that's not related to debugging.

Writing to a local file inside the installed extension directory is indeed a possibility but in addition to being slower than sending a Blob directly between contexts it would require either installing the extension inside the default downloads directory or to use a nativeMessaging app to write the file into the directory of an installed extension, finding which reliably and automatically is not always trivial since the browser may be launched with `--user-data-dir` parameter.

**Comment by guest271314** [comment]
@tophf I am curious what your proposed solution to this is?

**Comment by tophf** [comment]
Ideally `chrome` messaging should use the standard web platform messaging. The linked issue also mentions adding a `transfer` parameter to enable what the web platform's "transferrable" parameter does - at least in the non-broadcast methods i.e. ports.

**Comment by guest271314** [comment]
> Writing to a local file inside the installed extension directory is indeed a possibility but in addition to being slower than sending a Blob directly between contexts

Well, we cannot _transfer_ a `Blob` to different contexts at all.

The last time I tested `fetch()` is slower than File System Access API.

> require either installing the extension inside the default downloads directory

You can save the file anywhere you want excluding dot folders. You can write the file directly to the extension folder.

> Ideally `chrome` messaging should use the standard web platform messaging mechanism under the hood.

That is reasonable. Since we already have `"web_accessible_resources"` it _should be_ a simple matter to allow `BroadcastChannel` to read the manifest.json and allow messages, including transfers, to and from the `ServiceWorker`.

When WebRTC data channel is implemented in `Worker` contexts (only Safari implements that now) and eventually `ServiceWorker` contexts (not implemented by any browser) we can just use WebRTC data channel. However, Chrome still does not implement `Blob` for WebRTC data channels.

I would just open a dedicated chrome-extension: Tab, then transfer data from the Tab to the target Web page.

However, in practice in the field, I have _never_ encountered a Web site which removes the `iframe` I append to capture raw PCM. So the conjecture that a Web site will remove an `iframe` is just that, conjecture.

**Comment by guest271314** [comment]
In practice `"externally_connectable"` works. Just serialize the data to JSON and stream it. Done. I wouldn't be concerned about the given original `Blob`, `ArrayBuffer`, etc. I should point out here that `Blob` construction is expensive.

**Comment by guest271314** [comment]
In fact since we already have `"externally_connectable"` I see no reason why Chrome extensions cannot read the manifest.json, and set the site(s) listed as a `WindowClient` of the `ServiceWorker`, which will provide a means for direct communication to and from the `ServiceWorker` and the Web site using `postMessage()` and within `onfetch` use `respondWith()` when a `fetch()` request is made from the Web site listed in `"externally_connectable"`.

**Comment by guest271314** [comment]
@tophf The `iframe` approach works, though is not ideal. We are dealing with extensions here; we already have legacy `"externally_connectable"` and `"web_accessible_resources"`. It appears to me simply a matter of lack of will to update `"externally_connectable"` to provide the _option_ to, as you state, use Web platform `postMessage()` and `onmessage` from the sites the user explicitly lists in the manifest.json. It is certainly technically possible to do that.  That still won't let you _transfer_ a `Blob` though, as `Blob`s are not transferable.

**Comment by guest271314** [comment]
@tophf Alright, I modified this code https://github.com/guest271314/persistent-serviceworker/tree/main/chromium_extension_web_accessible_resources_iframe_message_event to remove the `iframe` after the `MessagePort` has been transferred to the Web page.

After a few tests, it takes half of 1 second for the `iframe` load` event to be dispatched and the `MessagePort` to be defined globally, if necessary.

```
iframe load event 0.4665
MessagePort from ServiceWorker defined globally 0.5065999999940395
```

```
iframe load event 0.40590000000596044
MessagePort from ServiceWorker defined globally 0.448
```

```
iframe load event 0.2025
MessagePort from ServiceWorker defined 0.21520000000298023
```

The first attempt to transfer 1GB of data crashed the Tab.

```
const buffer = new ArrayBuffer(2145386496 / 2);
```

Subsequent attempts to pass 500MB did not succeed

```
2145386496 / 4
```

Transferring one-quarter of 1GB did succeed.

In pertinent part

background.js

```
async function persistServiceWorker(src) {
  globalThis.port;
  return new Promise((resolve) => {
    onmessage = async (e) => {
      if (e.origin === new URL(src).origin) {
        if (e.ports.length) {
          globalThis.port = e.ports[0];
          console.log('MessagePort from ServiceWorker defined', (performance.now() - now) / 1000);
          port.onmessage = (e) => {
            console.log(e.data);
          };
          port.onmessageerror = (e) => {
            console.error(e);
          };
          if (iframe.parentNode) {
            // TODO: remove iframe when port is defined
            iframe.parentNode.removeChild(iframe);
          }
        } 
      }
    };
    let now = performance.now();
    iframe = document.createElement('iframe');
    iframe.style.display = 'none';
    iframe.name = location.origin;
    document.body.appendChild(iframe);
    iframe.onload = () => { console.log('iframe load event', (performance.now() - now) / 1000); resolve(); };
    iframe.src = `${src}index.html`;
  });
}
// ...
port1.onmessage = async (e) => {
  const buffer = new ArrayBuffer(2145386496 / 8); 
  const transfer = structuredClone(buffer, {transfer: [buffer]});
  port1.postMessage(transfer, [transfer]);
};
```

index.js

```
navigator.serviceWorker.onmessage = async (e) => {
  parent.postMessage(e.data, name, e.ports);
}
```

**Comment by guest271314** [comment]
The way I tested the code the click on extension icon does not immediately transfer the `ArrayBuffer`. If you want to send the `ArrayBuffer` at click on icon - which is not "instant" to transfer, move the `port1` message post to after `executeScript()` `Promise` is fulfilled

```
(await self.clients.matchAll({ includeUncontrolled: true }))[0]
.postMessage(null, [port2]);
const buffer = new ArrayBuffer(2145386496 / 8); 
const transfer = structuredClone(buffer, {transfer: [buffer]});
port1.postMessage(transfer, [transfer]);
```
Then it takes ~1.2-1.3 seconds for the entire process

```
iframe load event 0.1833999999910593
MessagePort from ServiceWorker defined 0.402
ArrayBuffer transferred from ServiceWorker 1.2995
```
```
iframe load event 0.2692000000029802
MessagePort from ServiceWorker defined 0.5351000000089406
ArrayBuffer(268173312)
ArrayBuffer transferred from ServiceWorker 1.3429000000059605
```

```
iframe load event 0.16240000000596047
MessagePort from ServiceWorker defined 0.17709999999403953
ArrayBuffer(268173312)
ArrayBuffer transferred from ServiceWorker 1.1602999999970198
```

**Comment by guest271314** [comment]
While the requirement is possible using an `ArrayBuffer` and an `iframe`, as a workaround for lack of a dedicated API to achieve the expecteded result, it is fragile for the reasons indicated by @tophf, namely use of `DOMNodeInserted` event and `MutationObserver` to remove the `iframe`.

**Comment by guest271314** [comment]
A couple more ways to do this.

background.js

```
self.addEventListener('activate', (event) => {
  event.waitUntil(clients.claim());
});
chrome.action.onClicked.addListener(async (tab) => {
  const [{ result }] = await chrome.scripting.executeScript({
    target: {
      tabId: tab.id,
      allFrames: true,
    },
    world: 'ISOLATED',
    args: [chrome.runtime.getURL('index.html') + '?' + new Date().getTime()],
    func: (src) => {
      const handleMessage = (e) => {
        if (src.includes(e.origin)) {
          console.log(new TextDecoder().decode(e.data));
        }
      };
      addEventListener('message', handleMessage, { once: true });
      open(
        src,
        location.href,
        `popup,left=0,top=0,width=1,height=1,location=0,resizable=1,scrollbars=0,toolbar=0,menubar=0,status=0,titlebar=0`
      );
      return true;
    },
  });
});

chrome.windows.onCreated.addListener(async (window) => {
  if (window.type === 'popup') {
    chrome.windows.update(window.id, {
      focused: false,
      top: 1,
      left: 1,
      height: 1,
      width: 1,
    });
  }
});

onfetch = async (e) => {
  console.log(e.request.url);
  if (e.request.url.includes('data')) {
    e.respondWith(new Response(new TextEncoder().encode('transfer')));
  }
};
```
script.js

```
const response = await fetch('./data');
const data = await response.arrayBuffer();
opener.postMessage(data, name, [data]);
close();
```

Unfortunately `setSelfAsOpener` is broken, so you'll get a flash of a 100x100 window with the above before `close()` is executed. If `setSelfAsOpener` worked in MV3 we could occlue the window altogether save for application tab which lists open programs. I don't think there is any way to reference the opened `window` from `windows.create()` from a Web page. I think this is a very simple model to use.

The below approach occuled the window opened with `windows.create()` and uses WebRTC data channel to communicate between the opened widown and `ServiceWorker` and opened window and arbitrary Web page. More moving parts.

background.js

```
addEventListener('activate', (event) => {
  event.waitUntil(clients.claim());
});

let resolve;
let promise = new Promise((_) => (resolve = _));
let local, remote;
let localTabId, remoteTabId;

globalThis.signaling = { remote: null, local: null };

chrome.action.onClicked.addListener(async (tab) => {
  localTabId = tab.id;
  let { tabs, id } = await chrome.windows.create({
    type: 'popup',
    focused: false,
    top: 1,
    left: 1,
    height: 1,
    width: 1,
    url: chrome.runtime.getURL('index.html'),
  });
  await chrome.windows.update(id, { focused: false });
 ({
    tabs: [{ id: remoteTabId }],
  } = await chrome.windows.get(id, { populate: true }));
});

let localResolve;
let localPromise = new Promise((_) => (localResolve = _));
let remoteResolve;
let remotePromise = new Promise((_) => (remoteResolve = _));
const handleUpdate = async (tabId, changeInfo, tab) => {
  if (
    changeInfo.status === 'complete' &&
    tab.url === chrome.runtime.getURL('index.html')
  ) {
    globalThis.signaling.remote = chrome.tabs.connect(tab.id, {
      name: 'remote',
    });
    globalThis.signaling.remote.onMessage.addListener(async (e) => {
      console.log(e);
      remoteResolve(e);
    });
    globalThis.signaling.local = chrome.tabs.connect(localTabId, {
      name: 'local',
    });
    globalThis.signaling.local.onMessage.addListener(async (e) => {
      if (e.local) {
        localResolve(e);
      }
      if (e === 'get-local') {
        const answer = await localPromise;
        globalThis.signaling.remote.postMessage(answer);
      }
      if (e === 'get-remote') {
        const answer = await remotePromise;
        console.log(answer);
        globalThis.signaling.local.postMessage(answer);
      }
    });
  }
};

chrome.tabs.onUpdated.addListener(handleUpdate);

onfetch = async (e) => {
  if (e.request.url.includes('data')) {
    e.respondWith(new Response(new TextEncoder().encode('transfer')));
  }
};
```

contentscript.js

```
(async () => {
  let resolve;
  let promise = new Promise((_) => (resolve = _));
  let port;
  chrome.runtime.onConnect.addListener(async (_) => {
    port = _;
    resolve();
  });
  await promise;
  const local = new RTCPeerConnection({
    sdpSemantics: 'unified-plan',
  });
  [
    'onsignalingstatechange',
    'oniceconnectionstatechange',
    'onicegatheringstatechange',
  ].forEach((event) => local.addEventListener(event, console.log));

  local.onicecandidate = async (event) => {
    if (!event.candidate) {
      if (local.localDescription.sdp.indexOf('a=end-of-candidates') === -1) {
        local.localDescription.sdp += 'a=end-of-candidates\r\n';
      }
      try {
        port.postMessage({
          local: true,
          sdp: local.localDescription.sdp,
        });
        await new Promise((resolve) => {
          const handleMessage = (message) => {
            port.onMessage.removeListener(handleMessage);
            resolve();
          };
          port.postMessage('get-local');
        });
      } catch (e) {
        console.error(e);
      }
    }
  };
  const channel = local.createDataChannel('transfer', {
    negotiated: true,
    id: 0,
  });
  channel.binaryType = 'arraybuffer';
  channel.onopen = async (e) => {
    console.log(e, local, channel);
  };
  channel.onmessage = async (e) => {
    console.log(new TextDecoder().decode(new Uint8Array(e.data)));
    local.close();
  };

  (async () => {
    try {
      const sdp = await new Promise((resolve) => {
        const handleMessage = (message) => {
          port.onMessage.removeListener(handleMessage);
          resolve(message.sdp);
        };
        port.onMessage.addListener(handleMessage);
        port.postMessage('get-remote');
      });
      await local.setRemoteDescription({ type: 'offer', sdp });
      const answer = await local.createAnswer();
      local.setLocalDescription(answer);
    } catch (e) {
      console.error(e);
    }
  })();
})();
```

**Comment by guest271314** [comment]
@tophf When we use `"run_at": "document_start"` and append the `iframe` to `document.documentElement` to transfer the `MessagePort` to the Web page from the `ServiceWorker` that script is run and completed before `load` event of `window` is fired, so no Web page scripts can remove the `iframe` because we will have already transferred the `MessagePort` to the Web page and removed the `iframe` programmatically before any Web page scripts have run.

**Comment by guest271314** [comment]
Now that offscreen documents are enabled by default we can use the WebRTC data channel approach, which should solve this

**Comment by guest271314** [comment]
This add the option to dynamically set one or more origin in `"matches"` array of `"externally_connectable"` in local manifest.json https://github.com/guest271314/offscreen-webrtc/blob/set-externally-connectable-persistent/README.md.

**Comment by guest271314** [comment]
@tophf @dotproto I think using `window.open()` is the most reliable way to transfer data from the `ServiceWorker` to the Web page https://github.com/guest271314/sw-transfer-stream as WebRTC https://github.com/guest271314/offscreen-webrtc does not function as expected when offline. Both solutions suffer from the need to create an entire HTML document just to transfer data.

I propose a way to dynamically set `Client` and `WindowClient` of both extension and non-extension `ServiceWorker`s here https://github.com/guest271314/requestClient/blob/main/README.md.
